File: algorithmic_toolbox/week5_dynamic_programming_1/4_longest_common_subsequence_of_two_sequences/lcs2.py
import sys
from typing import Sequence


def lcs2(a: Sequence, b: Sequence) -> int:  # noqa: C901
    """Longest common subsequence: insertion, deletion; no substitution"""

    if not set(a) & set(b):
        return 0

    m, n = len(a), len(b)
    dp = [[0 for _ in range(n + 1)] for _ in range(m + 1)]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    return dp[m][n]


if __name__ == "__main__":
    data = list(map(int, sys.stdin.read().split()))

    n = data[0]
    data = data[1:]
    a = data[:n]

    data = data[n:]
    m = data[0]
    data = data[1:]
    b = data[:m]

    print(lcs2(a, b))

File: algorithmic_toolbox/week5_dynamic_programming_1/4_longest_common_subsequence_of_two_sequences/lcs2.c
#include <stdio.h>
#include <stdlib.h>

int find_first_of(int* begin1, int* end1, int* begin2, int* end2) {
  while (begin1 != end1) {
    for (int* it = begin2; it != end2; it++)
      if (*it == *begin1) return 1;
    begin1++;
  }
  return 0;
}

/* Longest common subsequence: insertion, deletion; no substitution */
int lcs2(const int* a, const int a_size, const int* b, const int b_size) {
  int s[a_size];
  int t[b_size];
  for (int i = 0; i < a_size; i++) s[i] = a[i];
  for (int i = 0; i < b_size; i++) t[i] = b[i];

  int s_size = a_size;
  int t_size = b_size;

  int found = find_first_of(s, s + s_size, t, t + t_size);
  if (!found) return 0;

  int dp[a_size + 1][b_size + 1];

  for (int i = 0; i <= a_size; ++i)
    for (int j = 0; j <= b_size; ++j)
      dp[i][j] = 0;

  for (int i = 1; i <= a_size; ++i)
    for (int j = 1; j <= b_size; ++j)
      if (s[i - 1] == t[j - 1])
        dp[i][j] = dp[i - 1][j - 1] + 1;
      else
        dp[i][j] = dp[i - 1][j] > dp[i][j - 1] ? dp[i - 1][j] : dp[i][j - 1];

  return dp[a_size][b_size];
}

int main() {
  int n;
  scanf("%d", &n);
  int a[n];
  for (int i = 0; i < n; i++) scanf("%d", &a[i]);

  int m;
  scanf("%d", &m);
  int b[m];
  for (int i = 0; i < m; i++) scanf("%d", &b[i]);

  printf("%d\n", lcs2(a, n, b, m));

  return 0;
}
